url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/mario/engine/mapedit/LevelEditor.java

```diff
@@ -0,0 +1,151 @@
+    
+    private JCheckBox[] bitmapCheckboxes = new JCheckBox[8];
+
+    public LevelEditor()
+    {
+        super("Map Edit");
+        
+        try
+        {
+            System.out.println("System.getProperty(\"user.dir()\") = " + System.getProperty("user.dir"));
+            Level.loadBehaviors(new DataInputStream(LevelScene.class.getResourceAsStream("resources/tiles.dat")));
+//            Level.loadBehaviors(new DataInputStream(new FileInputStream("ch/idsia/mario/resources/tiles.dat")));
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+            JOptionPane.showMessageDialog(this, e.toString(), "Failed to load tile behaviors", JOptionPane.ERROR_MESSAGE);
+        }
+        
+        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
+        setSize(screenSize.width * 8 / 10, screenSize.height * 8 / 10);
+        setLocation((screenSize.width - getWidth()) / 2, (screenSize.height - getHeight()) / 2);
+        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+
+        tilePicker = new TilePicker();
+        JPanel tilePickerPanel = new JPanel(new BorderLayout());
+        tilePickerPanel.add(BorderLayout.WEST, tilePicker);
+        tilePickerPanel.add(BorderLayout.CENTER, buildBitmapPanel());
+        tilePickerPanel.setBorder(new TitledBorder(new EtchedBorder(), "Tile picker"));
+
+        JPanel lowerPanel = new JPanel(new BorderLayout());
+        lowerPanel.add(BorderLayout.EAST, tilePickerPanel);
+
+        JPanel borderPanel = new JPanel(new BorderLayout());
+        levelEditView = new LevelEditView(tilePicker);
+        borderPanel.add(BorderLayout.CENTER, new JScrollPane(levelEditView));
+        borderPanel.add(BorderLayout.SOUTH, lowerPanel);
+        borderPanel.add(BorderLayout.NORTH, buildButtonPanel());
+        setContentPane(borderPanel);
+
+        tilePicker.addTilePickChangedListener(this);
+    }
+
+    public JPanel buildBitmapPanel()
+    {
+        JPanel panel = new JPanel(new GridLayout(0, 1));
+        for (int i=0; i<8; i++)
+        {
+            bitmapCheckboxes[i] = new JCheckBox(Level.BIT_DESCRIPTIONS[i]);
+            panel.add(bitmapCheckboxes[i]);
+            if (Level.BIT_DESCRIPTIONS[i].startsWith("- ")) bitmapCheckboxes[i].setEnabled(false);
+            
+            final int id = i;
+            bitmapCheckboxes[i].addActionListener(new ActionListener()
+            {
+                public void actionPerformed(ActionEvent arg0)
+                {
+                    int bm = Level.TILE_BEHAVIORS[tilePicker.pickedTile&0xff]&0xff;
+                    bm&=255-(1<<id);
+                    if (bitmapCheckboxes[id].isSelected()) bm|=(1<<id);
+                    Level.TILE_BEHAVIORS[tilePicker.pickedTile&0xff] = (byte)bm;
+
+                    try
+                    {
+                        System.out.println("bm = " + bm);
+                        Level.saveBehaviors(new DataOutputStream(new FileOutputStream("SAVEDtiles.dat")));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *DataOutputStream*. Currently, there are execution paths that do not contain closure statements, for example, when Level.saveBehaviors throws an exception. Ensure the object returned by *DataOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).